lib/zip_fclose.c

```diff
@@ -41,8 +41,10 @@ ZIP_EXTERN int
 zip_fclose(zip_file_t *zf) {
     int ret;
 
-    if (zf->src)
+    if (zf->src) {
+        zip_source_close(zf->src);
```

Comment:
Ok, I can revert this change. I'll only point you to the commit message in https://github.com/nih-at/libzip/pull/286/commits/4a97e13e0b30e32ef4b61a67d3f21b9046a40cdf. Now this probably won't be a problem any more, because we always wrap the bottom-most source in some other sources, which happen to have ref count of 1, so they'll be closed on free.

I'd still say that `zip_source_free` is a misnomer (should be rather something like `zip_source_unref`), because it's freeing things only if ref count drops to 0. So if refcount is still higher, no free will happen and no closing either.

Reply:
You're right, we need to keep track which references called open the source. We will split the source struct in two parts, the actual source and a reference. But we can do this after the merge.